### PR TITLE
Rust: Implement ExecutionContext

### PR DESCRIPTION
**Pull request by jakelang**
Part of #168.

### CODE REVIEW

File: bindings/rust/evmc-vm/src/lib.rs [hunk 1]
```diff
@@ -18,6 +18,13 @@ pub struct ExecutionResult {
     create_address: ffi::evmc_address,
 }
 
+pub struct ExecutionContext<'a> {
+    rev: &'a ffi::evmc_revision,
+    msg: &'a ffi::evmc_message,
+    ctx: &'a mut ffi::evmc_context,
+    inst: &'a mut ffi::evmc_instance,
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: I'd prefer to use long names (or at least not 3-4 character abbreviations).

File: bindings/rust/evmc-vm/src/lib.rs [hunk 2]
```diff
@@ -61,6 +68,223 @@ impl ExecutionResult {
...
+        &self.msg.value
+    }
+
+    pub fn create2_salt(&self) -> &ffi::evmc_bytes32 {
+        &self.msg.create2_salt
+    }
+
+    pub fn account_exists(&mut self, addr: &ffi::evmc_address) -> bool {
+        unsafe {
+            assert!(!(*self.ctx.host).account_exists.is_some());
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: I think the assertions should take place in `new`.

However, in the long-run, no such assertions should be made, just catch panics in the macro and return an EVMC failure.

**Comment by axic** [comment]
Please add a commit making use of this in `example-rust-vm` for call kind.

Also please add some unit tests.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 3]
```diff
@@ -61,6 +68,223 @@ impl ExecutionResult {
...
+                addr as *const ffi::evmc_address,
+                data as *const [u8] as *const u8,
+                data.len(),
+                topics as *const [ffi::evmc_bytes32] as *const ffi::evmc_bytes32,
+                topics.len(),
+            )
+        }
+    }
+
+    pub fn get_instance(&self) -> &ffi::evmc_instance {
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Is this only needed by the macro stuff? If so I'd remove it from this PR.

File: bindings/rust/evmc-vm/src/lib.rs (hunk 1 above)
Comment (jakelang) [reply bindings/rust/evmc-vm/src/lib.rs]: i plan to scrap these members anyway in favor of a higher level representation of all the context data. will keep it in mind.

File: bindings/rust/evmc-vm/src/lib.rs (hunk 2 above)
Comment (jakelang) [reply bindings/rust/evmc-vm/src/lib.rs]: this makes sense.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 4]
```diff
@@ -61,6 +68,223 @@ impl ExecutionResult {
...
+
+    pub fn destination(&self) -> &ffi::evmc_address {
+        &self.msg.destination
+    }
+
+    pub fn sender(&self) -> &ffi::evmc_address {
+        &self.msg.sender
+    }
+
+    pub fn input_data(&self) -> Option<Vec<u8>> {
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Can this be made into a slice to the input?

File: bindings/rust/evmc-vm/src/lib.rs [hunk 5]
```diff
@@ -61,6 +67,216 @@ impl ExecutionResult {
...
+        &self.message.destination
+    }
+
+    pub fn sender(&self) -> &ffi::evmc_address {
+        &self.message.sender
+    }
+
+    pub fn input_data(&self) -> Option<Vec<u8>> {
+        if !self.message.input_data.is_null() {
+            let mut ret: Vec<u8> = Vec::with_capacity(self.message.input_size);
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Need to also use `set_len`.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 6]
```diff
@@ -61,6 +67,219 @@ impl ExecutionResult {
...
+            tx_context: _tx_context,
+        }
+    }
+
+    pub fn call_kind(&self) -> &ffi::evmc_call_kind {
+        &self.message.kind
+    }
+
+    pub fn is_static(&self) -> bool {
+        self.message.flags == 1
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Not sure we want this helper, but in any case this should use `ffi::EVMC_STATIC` or `ffii::evmc_flags::EVMC_STATIC`.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 7]
```diff
@@ -18,6 +18,12 @@ pub struct ExecutionResult {
     create_address: ffi::evmc_address,
 }
 
+pub struct ExecutionContext<'a> {
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Rustdoc?

File: bindings/rust/evmc-vm/src/lib.rs [hunk 8]
```diff
@@ -308,4 +480,75 @@ mod tests {
             }
         }
     }
+
+    unsafe extern "C" fn get_dummy_txctx(context: *mut ffi::evmc_context) -> ffi::evmc_tx_context {
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Any reason for the shorthand `txctx`? :)

Comment (jakelang) [reply bindings/rust/evmc-vm/src/lib.rs]: its a test helper leave me alone :(
will fix it in a bit

Comment (axic) [reply bindings/rust/evmc-vm/src/lib.rs]: Can we name this nicely?

File: bindings/rust/evmc-vm/src/lib.rs [hunk 9]
```diff
@@ -308,4 +480,75 @@ mod tests {
...
+            sender: ffi::evmc_address { bytes: [0u8; 20] },
+            input_data: std::ptr::null() as *const u8,
+            input_size: 0,
+            value: ffi::evmc_uint256be { bytes: [0u8; 32] },
+            create2_salt: ffi::evmc_uint256be { bytes: [0u8; 32] },
+        }
+    }
+
+    #[test]
+    fn execution_context() {
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Please test at last one function, for example `get_code_size`.

**Comment by axic** [comment]
Can we squash some commits or at least name them properly?

File: bindings/rust/evmc-vm/src/lib.rs (hunk 9 above)
Comment (jakelang) [reply bindings/rust/evmc-vm/src/lib.rs]: Done in a separate test case.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 10]
```diff
@@ -308,4 +480,99 @@ mod tests {
...
+            block_gas_limit: 105023,
+            block_difficulty: ffi::evmc_uint256be { bytes: [0xaa; 32] },
+        }
+    }
+
+    unsafe extern "C" fn get_dummy_code_size(
+        _context: *mut ffi::evmc_context,
+        _addr: *const ffi::evmc_address,
+    ) -> usize {
+        105023usize
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Nitpicking but can you use `105023 as usize`?

Never liked these compound forms in C++ either.

File: bindings/rust/evmc-vm/src/lib.rs [hunk 11]
```diff
@@ -61,6 +69,170 @@ impl ExecutionResult {
...
+        addr: &ffi::evmc_address,
+        data: &[u8],
+        topics: &[ffi::evmc_bytes32],
+    ) {
+        unsafe {
+            assert!((*self.context.host).emit_log.is_some());
+            (*self.context.host).emit_log.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                addr as *const ffi::evmc_address,
+                data as *const [u8] as *const u8,
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Can you avoid this cast?

File: bindings/rust/evmc-vm/src/lib.rs [hunk 12]
```diff
@@ -61,6 +69,170 @@ impl ExecutionResult {
...
+                data.len(),
+                topics as *const [ffi::evmc_bytes32] as *const ffi::evmc_bytes32,
```
Comment (axic) [inline bindings/rust/evmc-vm/src/lib.rs]: Not sure about alignment here and this looks nasty. At the bare minimum with need a FIXME note and explanation.

**Pull request merged**